src/pybind/mgr/dashboard/controllers/rbd.py

```diff
@@ -1,136 +1,431 @@
+
+
+def _sort_features(features, enable=True):
+    """
+    Sorts image features according to feature dependencies:
+
+    object-map depends on exclusive-lock
+    journaling depends on exclusive-lock
+    fast-diff depends on object-map
+    """
+    ORDER = ['exclusive-lock', 'journaling', 'object-map', 'fast-diff']
+
+    def key_func(feat):
+        try:
+            return ORDER.index(feat)
+        except ValueError:
+            return id(feat)
+
+    features.sort(key=key_func, reverse=not enable)
+
+
+@ApiController('block/image')
 @AuthRequired()
 class Rbd(RESTController):
 
-    RBD_FEATURES_NAME_MAPPING = {
-        rbd.RBD_FEATURE_LAYERING: "layering",
-        rbd.RBD_FEATURE_STRIPINGV2: "striping",
-        rbd.RBD_FEATURE_EXCLUSIVE_LOCK: "exclusive-lock",
-        rbd.RBD_FEATURE_OBJECT_MAP: "object-map",
-        rbd.RBD_FEATURE_FAST_DIFF: "fast-diff",
-        rbd.RBD_FEATURE_DEEP_FLATTEN: "deep-flatten",
-        rbd.RBD_FEATURE_JOURNALING: "journaling",
-        rbd.RBD_FEATURE_DATA_POOL: "data-pool",
-        rbd.RBD_FEATURE_OPERATIONS: "operations",
-    }
-
-    def __init__(self):
-        super(Rbd, self).__init__()
-        self.rbd = None
-
-    @staticmethod
-    def _format_bitmask(features):
-        """
-        Formats the bitmask:
-
-        >>> Rbd._format_bitmask(45)
-        ['deep-flatten', 'exclusive-lock', 'layering', 'object-map']
-        """
-        names = [val for key, val in Rbd.RBD_FEATURES_NAME_MAPPING.items()
-                 if key & features == key]
-        return sorted(names)
-
-    @staticmethod
-    def _format_features(features):
-        """
-        Converts the features list to bitmask:
-
-        >>> Rbd._format_features(['deep-flatten', 'exclusive-lock', 'layering', 'object-map'])
-        45
-
-        >>> Rbd._format_features(None) is None
-        True
-
-        >>> Rbd._format_features('not a list') is None
-        True
-        """
-        if not features or not isinstance(features, list):
-            return None
-
-        res = 0
-        for key, value in Rbd.RBD_FEATURES_NAME_MAPPING.items():
-            if value in features:
-                res = key | res
-        return res
+    # set of image features that can be enable on existing images
+    ALLOW_ENABLE_FEATURES = set(["exclusive-lock", "object-map", "fast-diff",
+                                 "journaling"])
 
-    @ViewCache()
-    def _rbd_list(self, pool_name):
-        ioctx = mgr.rados.open_ioctx(pool_name)
-        self.rbd = rbd.RBD()
-        names = self.rbd.list(ioctx)
-        result = []
-        for name in names:
-            i = rbd.Image(ioctx, name)
-            stat = i.stat()
-            stat['name'] = name
-            stat['id'] = i.id()
-            features = i.features()
+    # set of image features that can be disabled on existing images
+    ALLOW_DISABLE_FEATURES = set(["exclusive-lock", "object-map", "fast-diff",
+                                  "deep-flatten", "journaling"])
```

Comment:
How can we make sure, this is not getting out-of-date?

Reply:
Update the code when someone changes the Ceph RBD component. :-)

Reply:
We can expose the RBD_FEATURES_MUTABLE bitmasks via the RBD python module